lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,183 @@
+use crate::encoding::BuildError;
+use bytes::{BufMut, BytesMut};
+use lookup::lookup_v2::ConfigOwnedTargetPath;
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerConfig {
+    /// The CSV Serializer Options.
+    pub csv: CsvSerializerOptions,
+}
+
+impl CsvSerializerConfig {
+    /// Creates a new `CsvSerializerConfig`.
+    pub const fn new(csv: CsvSerializerOptions) -> Self {
+        Self { csv }
+    }
+
+    /// Build the `CsvSerializer` from this configuration.
+    pub fn build(&self) -> Result<CsvSerializer, BuildError> {
+        if self.csv.fields.is_empty() {
+            Err("At least one CSV field must be specified".into())
+        } else {
+            Ok(CsvSerializer::new(self.csv.fields.clone()))
+        }
+    }
+
+    /// The data type of events that are accepted by `CsvSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // CSV, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerOptions {
+    /// The CSV Serializer fields.
+    ///
+    /// The order of the fields determines the order of the columns in the CSV output.
+    /// Make sure to use the same order when parsing the CSV output.
+    ///
+    /// If a field is not present in the event, the output will be `NaN`.
+    ///
+    /// Some value types: Array, Regex, Object are not supported by the CSV format.
+    /// If a field is of an unsupported type, the output will be `NaN`.
+    pub fields: Vec<ConfigOwnedTargetPath>,
```

Comment:
We're still discussing this compared to the `only_fields` option, and how that interacts with the `csv` codec's ordering requirements. I think this is the primary blocker on this PR as it stands.

Reply:
Actually, less discussion than I expected - we're happy to move forward with the existing implementation 😅